services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java

```diff
@@ -0,0 +1,281 @@
+                  veniceParentHelixAdmin.getCurrentVersionsForMultiColos(cluster, storeName);
+              Set<String> remainingRegions = getRegionsForVersionSwap(coloToVersions, targetRegions);
+
+              StoreResponse targetRegionStoreResponse =
+                  getStoreForRegion(cluster, getTargetRegion(targetRegions), storeName);
+              if (targetRegionStoreResponse.isError()) {
+                LOGGER.warn("Got error when fetching targetRegionStore: {}", targetRegionStoreResponse.getError());
+                continue;
+              }
+
+              StoreInfo targetRegionStore = targetRegionStoreResponse.getStore();
+              Optional<Version> version = targetRegionStore.getVersion(targetVersionNum);
+              if (!version.isPresent()) {
+                LOGGER.warn(
+                    "Unable to find version {} for store: {} in regions: {}",
+                    targetVersionNum,
+                    storeName,
+                    store.getTargetSwapRegion());
+                continue;
+              }
+
+              // Do not perform version swap for davinci stores
+              // TODO remove this check once DVC delayed ingestion is completed
+              if (version.get().getIsDavinciHeartbeatReported()) {
+                LOGGER.info(
+                    "Skipping version swap for store: {} on version: {} as it is davinci",
+                    storeName,
+                    targetVersionNum);
+                continue;
+              }
+
+              // Check that push is completed in target regions
+              String kafkaTopicName = Version.composeKafkaTopic(storeName, targetVersionNum);
+              Admin.OfflinePushStatusInfo pushStatusInfo =
+                  veniceParentHelixAdmin.getOffLinePushStatus(cluster, kafkaTopicName);
+              boolean didPushCompleteInTargetRegions = true;
+              for (String targetRegion: targetRegions) {
+                String executionStatus = pushStatusInfo.getExtraInfo().get(targetRegion);
+                if (!executionStatus.equals(ExecutionStatus.COMPLETED.toString())) {
+                  didPushCompleteInTargetRegions = false;
+                  LOGGER.warn(
+                      "Skipping version swap for store: {} on version: {} as push is not complete in target region {}",
+                      storeName,
+                      targetVersionNum,
+                      targetRegion);
+                }
+              }
+
+              if (!didPushCompleteInTargetRegions) {
+                continue;
+              }
+
+              // Check that push is complete in non target regions
+              int numNonTargetRegionsCompleted = remainingRegions.size();
+              for (String remainingRegion: remainingRegions) {
+                String executionStatus = pushStatusInfo.getExtraInfo().get(remainingRegion);
+                if (!executionStatus.equals(ExecutionStatus.COMPLETED.toString())) {
```

Comment:
can you explicitly check for error status, they could be in other non-terminal states,